src/sign/keyset.rs

```diff
@@ -0,0 +1,1906 @@
+                        inner_k.timestamps.published = Some(now.clone());
+                        continue 'outer;
+                    }
+                    KeyType::Csk(
+                        ref mut ksk_keystate,
+                        ref mut zsk_keystate,
+                    ) => {
+                        if *ksk_keystate
+                            != (KeyState {
+                                old: false,
+                                signer: false,
+                                present: false,
+                                at_parent: false,
+                            })
+                        {
+                            return Err(Error::WrongKeyState);
+                        }
+
+                        // Move key state to Active.
+                        ksk_keystate.present = true;
+                        ksk_keystate.signer = true;
+
+                        if *zsk_keystate
+                            != (KeyState {
+                                old: false,
+                                signer: false,
+                                present: false,
+                                at_parent: false,
+                            })
+                        {
+                            return Err(Error::WrongKeyState);
+                        }
+
+                        // Move key state to Incoming.
+                        zsk_keystate.present = true;
+
+                        inner_k.timestamps.published = Some(now.clone());
+                        continue 'outer;
+                    }
+                    _ => {
+                        return Err(Error::WrongKeyType);
+                    }
+                }
+            }
+
+            return Err(Error::KeyNotFound);
+        }
+
+        // Make sure we have at least one KSK key in incoming state.
+        if !keys.iter_mut().any(|k| match &k.keytype {
+            KeyType::Ksk(keystate) | KeyType::Csk(keystate, _) => {
+                !keystate.old && keystate.present
+            }
+            _ => false,
+        }) {
+            return Err(Error::NoSuitableKeyPresent);
+        }
+        // Make sure we have at least one ZSK key in incoming state.
+        if !keys.iter_mut().any(|k| match &k.keytype {
+            KeyType::Zsk(keystate) | KeyType::Csk(_, keystate) => {
+                !keystate.old && keystate.present
+            }
+            _ => false,
+        }) {
+            return Err(Error::NoSuitableKeyPresent);
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+/// The state of a single key.
+///
+/// The state includes a way to refer to the public key and optionally a
+/// way to refer to the provate key. The state includes the type of the
+/// key (which in itself includes the key state) and a list of timestamps
+/// that mark the various stages in the life of a key.
+pub struct Key {
+    pubref: String,
+    privref: Option<String>,
```

Comment:
I think we need a better way to allow users to insert custom per-key information, because it might not always be a string.  I'm not sure `privref` is all they'll need.

Reply:
We have to see how this evolves when we start writing signers. Though my current line of thinking is to get rid of privref and let the user manage additional state keyed on pubref.

Reply:
Yeah, I like that idea!